lib/bap_service/bap_service.ml

```diff
@@ -0,0 +1,109 @@
+open Core_kernel.Std
+open Bap_future.Std
+open Regular.Std
+
+type void
+type literal = (void,void,void) format
+
+type 'a named = {
+  name : string;
+  desc : string;
+  data : 'a;
+} [@@deriving bin_io, compare, fields, sexp]
+
+type service = string
+[@@deriving bin_io, compare, sexp]
+
+type provider = service named
+[@@deriving bin_io, compare, sexp]
+
+type product = {
+  digest   : string;
+  provider : provider;
+} [@@deriving bin_io, compare, sexp]
+
+module Table = String.Table
+
+module Service = struct
+  type t = service
+
+  type info = {
+    products  : product stream;
+    update    : product signal;
+    providers : provider Table.t;
+  } [@@deriving fields]
+
+  let all : info named Table.t = Table.create ()
+
+  let info s = data (Hashtbl.find_exn all s)
+  let products  s = products (info s)
+  let update    s = update (info s)
+  let providers s = providers (info s) |> Hashtbl.data
+  let request = products
+
+  let add_provider serv prov =
+    Hashtbl.set (info serv).providers prov.name prov
+
+  let check_uuid x =
+    if Option.is_none (Uuidm.of_string x) then
+      invalid_arg "Invalid UUID format";
+    if Hashtbl.mem all x then
+      invalid_argf "UUID %s is already in use" x ()
```

Comment:
Please provide more information to a user. Suppose you are the user, who has no idea about those services an you get the following error:
```
$ bap --version
Program terminates: UUID 6a0037e5-5bc9-4bfd-a6a6-a495fc4387fa is already in use.
```

By whom, what for, why is it bad. 

The right error message would be:

```
Unable to register the `XXX` service as there is a service with the same UUID.
Please provide a unique service identifier. 
```